src/shared/clock-util.c

```diff
@@ -133,21 +133,27 @@ int clock_reset_timewarp(void) {
 }
 
 int clock_apply_epoch(ClockChangeDirection *ret_attempted_change) {
-        usec_t epoch_usec, now_usec;
+        usec_t epoch_usec = 0, timesyncd_usec = 0, now_usec;
         struct stat st;
 
         /* NB: we update *ret_attempted_change in *all* cases, both
          * on success and failure, to indicate what we intended to do! */
 
         assert(ret_attempted_change);
 
-        if (stat(EPOCH_CLOCK_FILE, &st) < 0) {
-                if (errno != ENOENT)
-                        log_warning_errno(errno, "Cannot stat " EPOCH_CLOCK_FILE ": %m");
-
-                epoch_usec = (usec_t) TIME_EPOCH * USEC_PER_SEC;
-        } else
+        if (stat(TIMESYNCD_CLOCK_FILE, &st) >= 0)
                 epoch_usec = timespec_load(&st.st_mtim);
+        else if (errno != ENOENT)
+                log_warning_errno(errno, "Could not stat %s, ignoring: %m", TIMESYNCD_CLOCK_FILE);
+
+        if (stat(EPOCH_CLOCK_FILE, &st) >= 0)
+                timesyncd_usec = timespec_load(&st.st_mtim);
+        else if (errno != ENOENT)
+                log_warning_errno(errno, "Could not stat %s, ignoring: %m", EPOCH_CLOCK_FILE);
+
+        epoch_usec = MAX3(epoch_usec,
+                          timesyncd_usec,
+                          (usec_t) TIME_EPOCH * USEC_PER_SEC);
```

Comment:
if TIME_EPOCH is zero, and the files don't exist we'll actually reset the clock to zero even if valid. That's a bit weird.

Maybe just explicitly check for a zero epoch_usec below and turn it into CLOCK_CHANGE_NOOP

Reply:
It's a bit of a corner case, but people occasionally do explicitly set mtimes to 0, so TIME_EPOCH could realistically be 0. I added an explicit check.